components/expression/expression-loader.js

```diff
@@ -52,117 +54,237 @@ genotet.ExpressionLoader.prototype.loadExpressionMatrix_ = function(matrixName,
+        matrixname: originParams.matrixName,
+        geneRegex: regex,
+        conditionRegex: conditionRegex
+      });
+      this.data.matrix = data;
+      this.signal('loadComplete');
+      this.signal('updatePanel');
+    }.bind(this), 'jsonp')
+    .fail(this.fail.bind(this, 'cannot update genes in the expression', params));
 };
 
-LoaderHeatmap.prototype.loadData = function(identifier) {
-  this.parentView.layout.showMsg('Loading...');
-  this.parentView.viewdata = {};
-  this.parentView.viewdata.lineData = new Array();
-  this.lastIdentifier = identifier;
-  if (identifier.name != null && identifier.name != '') {
-    this.loadLine(identifier.mat, identifier.name);
+/**
+ * Updates the conditions in the current expression.
+ * @param {string} method Update method, either 'set' or 'add'.
+ * @param {string} conditionRegex Regex that selects the conditions to be updated.
+ */
+genotet.ExpressionLoader.prototype.updateConditions = function(method, conditionRegex, originParams) {
+  var regex = this.data.matrix.conditionRegex;
+  conditionRegex = conditionRegex.toUpperCase();
+  switch (method) {
+    case 'setCondition':
+      // Totally replace the regex.
+      regex = conditionRegex;
+      this.data.conditions = [];
+      this.data.conditions.push(conditionRegex);
+      break;
+    case 'addCondition':
+      // Concat the two regex's. We need to include the previously existing
+      // genes too so as to find the edges between the new genes and old
+      // ones.
+      regex += '|' + conditionRegex;
+      this.data.conditions.push(conditionRegex);
+      break;
+    case 'removeCondition':
+      // Remove the regex.
+      var index = this.data.conditions.indexOf(conditionRegex);
+      if (index == -1) {
+        genotet.error('invalid condition regex', conditionRegex);
+        return;
+      };
+      this.data.conditions.splice(index, 1);
+      regex = '';
+      this.data.conditions.forEach(function(conditionRegexs, i) {
+        regex += conditionRegexs + (i == this.data.conditions.length - 1 ? '' : '|');
+      }, this);
+      break;
   }
-  this.loadHeatmap(identifier.mat, identifier.exprows, identifier.expcols);
+  var geneRegex = '';
+  this.data.genes.forEach(function(geneNames, i) {
+    geneRegex += geneNames + (i == this.data.genes.length - 1 ? '' : '|');
+  }, this);
+  this.signal('loadStart');
+  var params = {
+    type: 'expression',
+    mat: originParams.matrixName,
+    exprows: geneRegex,
+    expcols: regex
+  };
+  $.get(genotet.data.serverURL, params, function(data) {
+      // Store the last applied data selectors.
+      _(data).extend({
+        matrixname: originParams.matrixName,
+        geneRegex: geneRegex,
+        conditionRegex: regex
+      });
+      this.data.matrix = data;
+      this.signal('loadComplete');
+    }.bind(this), 'jsonp')
+    .fail(this.fail.bind(this, 'cannot update conditions in the expression', params));
 };
 
-LoaderHeatmap.prototype.loadComplete = function() {
```

Comment:
Please remove all the old codes below. It doens't seem that they will be re-used. Thanks.